patches/server/0066-Add-Velocity-natives-for-encryption-and-compression.patch

```diff
@@ -0,0 +1,313 @@
++                }
++//                byte[] abyte = new byte[packetdataserializer.readableBytes()];
++//
++//                packetdataserializer.readBytes(abyte);
++//                this.a.setInput(abyte);
++//                byte[] abyte1 = new byte[i];
++//
++//                this.a.inflate(abyte1);
++//                list.add(Unpooled.wrappedBuffer(abyte1));
++//                this.a.reset();
++                // Tuinity end
+             }
+ 
+         }
+diff --git a/src/main/java/net/minecraft/server/PacketDecrypter.java b/src/main/java/net/minecraft/server/PacketDecrypter.java
+index c85f291c5b22a8e85c7556b220cba698701748f2..771cc0f4fa98be294abba65c83442205b6b0ef0b 100644
+--- a/src/main/java/net/minecraft/server/PacketDecrypter.java
++++ b/src/main/java/net/minecraft/server/PacketDecrypter.java
+@@ -8,13 +8,24 @@ import javax.crypto.Cipher;
+ 
+ public class PacketDecrypter extends MessageToMessageDecoder<ByteBuf> {
+ 
+-    private final PacketEncryptionHandler a;
++    // Tuinity start
++    private final com.velocitypowered.natives.encryption.VelocityCipher cipher;
++    //private final PacketEncryptionHandler a;
+ 
+-    public PacketDecrypter(Cipher cipher) {
+-        this.a = new PacketEncryptionHandler(cipher);
++    public PacketDecrypter(javax.crypto.SecretKey key /* Cipher cipher */) throws java.security.GeneralSecurityException {
++        //this.a = new PacketEncryptionHandler(cipher);
++        this.cipher = com.velocitypowered.natives.util.Natives.cipher.get().forDecryption(key);
+     }
+ 
+     protected void decode(ChannelHandlerContext channelhandlercontext, ByteBuf bytebuf, List<Object> list) throws Exception {
+-        list.add(this.a.a(channelhandlercontext, bytebuf));
++        ByteBuf compatible = com.velocitypowered.natives.util.MoreByteBufUtils.ensureCompatible(channelhandlercontext.alloc(), cipher, bytebuf).slice();
++        try {
++            cipher.process(compatible);
++            list.add(compatible);
++        } catch (Exception e) {
++            compatible.release(); // compatible will never be used if we throw an exception
++            throw e;
++        }
+     }
++    // Tuinity end
+ }
+diff --git a/src/main/java/net/minecraft/server/PacketEncrypter.java b/src/main/java/net/minecraft/server/PacketEncrypter.java
+index e35369476839e9622520af1027d7478aa6d1b037..aba14794cc4cb114fea17bb92816ac29a64b44f8 100644
+--- a/src/main/java/net/minecraft/server/PacketEncrypter.java
++++ b/src/main/java/net/minecraft/server/PacketEncrypter.java
+@@ -5,15 +5,38 @@ import io.netty.channel.ChannelHandlerContext;
+ import io.netty.handler.codec.MessageToByteEncoder;
+ import javax.crypto.Cipher;
+ 
+-public class PacketEncrypter extends MessageToByteEncoder<ByteBuf> {
++// Tuinity start
++// We rewrite this class as the Velocity natives support in-place encryption
++import io.netty.handler.codec.MessageToMessageEncoder; // An unfortunate import, but this is required to fix a compiler error
```

Comment:
Explanation: At least in 1.16.3, Spigot mappings do not rename the obfuscated class `io`.

Reply:
yeah they do the same thing with the `it` class so we gotta import fastutil as well
this is exactly why md_5 needs to name classes, among other decompile issues